Docs: Add Datadog guide. (#21950)
gravitational/teleport: docs/pages/management/export-audit-events/datadog.mdx

---
title: Export Teleport Audit Events with Datadog
description: How to configure the Teleport Event Handler plugin and Fluentd to send audit logs to Datadog
---

Datadog is a SAAS monitoring and security platform. In this guide, we'll explain
how to forward Teleport audit events to Datadog using Fluentd.

The Teleport Event Handler is designed to communicate with Fluentd using mTLS

> to establish a secure channel. While we can use this same configuration to connect

Author: In those setups we connect Event Handler to _some_ tool that's part of those ecosystems, as opposed to a third-party like Fluentd.

> to establish a secure channel. While we can use this same configuration to connect
> directly to [Splunk](./splunk.mdx) or [Elastic](./elastic-stack.mdx), Datadog uses
> an API endpoint with keys, so we will use Fluentd as an intermediary.

Reviewer: ```suggestion
to establish a secure channel. In this setup, the Event Handler sends events to Fluentd, which forwards them to Datadog using an API key to authenticate.
```
How about something like this? I don't think we need to know how the Splunk/Elastic setups work in order to use this guide, and this describes the way Fluentd works here.

{/*
The Graph below is made with this mermaidjs logic:
graph LR
style e fill: #a379ff,color: #ffffff
style f fill: #a379ff,color: #ffffff
style c fill: #a379ff,color: #ffffff
style g fill: #a379ff,color: #ffffff
style i fill: #a379ff,color: #ffffff
style e fill: #844cff,color: #ffffff
style monitor fill:#a8afb1,stroke: #000000

a(User)
a-->e

subgraph local [Local Network]
    style local fill:#C0C0C0,stroke: #000000
    direction LR
    subgraph teleport [ ]
        direction TB
        style teleport fill:none,stroke:none
            e(Teleport Proxy)
        e<-->f(Teleport Auth)
        e<-->g(Teleport Service)
        e<-->i(Teleport Service)
        g<-->j([Infrastructure])
        i<-->k([Infrastructure])
    end

    subgraph monitor [Monitoring]
        c(teleport-event-handler)
        d(fluentd)
    end
    c-->d
end
e-->c
d-->h(Datadog)
*/}

![Flowchart of Teleport with events forwarded to Datadog](../../../img/management/datadog-flowchart.png)

> ## Prereqs

Author: Oops, should have caught that myself after my first writing run-through.

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- A [Datadog](https://www.datadoghq.com/) account.
- A server, virtual machine, Kubernets cluster, or Docker environment to run the
  Event Handler. The instructions below assume a local Docker container for testing.
- Fluentd version v(=fluentd.version=) or greater. The Teleport Event Handler
  will create a new `fluent.conf` file you can integrate into an existing Fluentd
  system, or use with a fresh setup. The instructions below assume a local Docker
  container for testing.

(!docs/pages/includes/tctl.mdx!)

## Step 1/6. Install the Event Handler plugin

The Teleport event handler runs alongside the Fluentd forwarder, receives events
from Teleport's events API, and forwards them to Fluentd.

(!docs/pages/includes/install-event-handler.mdx!)

## Step 2/6. Configure the plugin

> (!docs/pages/includes/configure-event-handler.mdx!)

Author: Added a docker tab to the configure partial.

> (!docs/pages/includes/configure-event-handler.mdx!)

Author: ....and removed it, because it seems that you can't run the configure command from outside the container.

> (!docs/pages/includes/configure-event-handler.mdx!)

Reviewer: 👍 Would it make sense to change the "The instructions below assume a local Docker container for testing" note in the Prerequisites to, "The instructions below assume that you are running the Event Handler on your workstation" or something like that, so we're not tied to the Docker approach?

```code
$ tctl auth sign --user=teleport-event-handler --out=identity
```

This command creates one PEM-encoded file, `identity`. The identity file
includes both TLS and SSH credentials. The Event Handler plugin uses the SSH
credentials to connect to the Proxy Service, which establishes a reverse tunnel
connection to the Auth Service. The plugin uses this reverse tunnel, along with
your TLS credentials, to connect to the Auth Service's gRPC endpoint.

</TabItem>
<TabItem label="Helm Chart" scope={["cloud"]}>

If you are planning to use the Helm Chart, you'll need to generate the keys
with the `file` format, then create a secret in Kubernetes.

Create the identity using the following command:

```code
$ tctl auth sign --format=file --user=teleport-event-handler --out=identity
```

Then create the Kubernetes secret:

```code
$ kubectl create secret generic teleport-event-handler-identity --from-file=auth_id=identity
```

These commands should result in a PEM-encoded file, `identity`, and a secret
in Kubernetes with the name `teleport-event-handler-identity`.

</TabItem>
</Tabs>

## Step 5/6. Install Fluentd output plugin for Datadog

> In order for Fluentd to communicate with Datadog, it requires the [Fluentd output

Author: I'd like to push back on this a bit. I know what our style guide says, but putting myself in the position of a reader, I'd like an opportunity to review a piece of third party software instead of blindly installing in into my system.

Also consider the future where the instructions below might go stale, and until we know/resolve it the first party docs should (finger crossed) update with changes.

> In order for Fluentd to communicate with Datadog, it requires the [Fluentd output

Reviewer: In that case, would it make sense to tell the reader to follow the installation instructions in the `fluent-plugin-datadog` README?

> In order for Fluentd to communicate with Datadog, it requires the [Fluentd output

Author: Not for the other reader who _doesn't_ want to review their documentation, or already has and is following this guide to recreate on another environment, etc. This is a middle-ground where the user can _choose_ to follow the link but doesn't have to, and our docs contain a reasonable amount of content for a third party system.

plugin for Datadog](https://github.com/DataDog/fluent-plugin-datadog). Install
the plugin on your Fluentd host using either `gem` or the `td-agent`, if installed:

```code
# Using Gem
$ gem install fluent-plugin-datadog

# Using td-agent
$ /usr/sbin/td-agent-gem install fluent-plugin-datadog
```

<Admonition type="tip" title="Testing Locally?">

If you're running Fluentd in a local Docker container for testing, you can adjust
the entrypoint to an interactive shell as the root user, so you can install the
plugin before starting Fluentd:

```code

> $ docker run -u $(id -u root):$(id -g root) -p 8888:8888 -v $(pwd):/keys -v \

Author: The current working directory is where the certs are when you run the `configure` command. Creating a specific directory means additional host-side instructions that may or may not be relevant to the end user. Mounting each key individually adds a lot of complexity to the commands.

These are examples. at some point we have to trust the reader to know enough about their own environment and common tools like Docker to be able to adjust as needed for themselves.

    certPath: "client.crt"
    keyPath: "client.key"

persistentVolumeClaim:
  enabled: true
```

To start the event handler in Kubernetes, run the following command:

```code
$ helm install teleport-plugin-event-handler teleport/teleport-plugin-event-handler \
  --values teleport-plugin-event-handler-values.yaml \
  --version (=teleport.plugin.version=)
```

</TabItem>
</Tabs>

<Admonition type="note">

This example will start exporting from `May 5th 2021`:

```code
$ ./teleport-event-handler start --config teleport-event-handler.toml --start-time "2022-02-02T00:00:00Z"
```

The start time can be set only once, on the first run of the tool.

If you want to change the time frame later, remove the plugin state directory
that you specified in the `storage` field of the handler's configuration file.

</Admonition>

Once the handler starts, you will see notifications in Fluentd about scanned and forwarded events:

```txt
INFO[0046] Event sent id=0b5f2a3e-faa5-4d77-ab6e-362bca0994fc ts="2021-06-08 11:00:56.034 +0000 UTC" type=user.login
...
```

The Logs view in Datadog should now report your Teleport cluster events:

![Datadog Logs](../../../img/management/datadog-logs.png)

> Review the Fluentd output plugin for Datadog [README

Author: Moving.

---
title: Monitor Teleport Audit Events with Datadog
description: How to configure the Teleport Event Handler plugin and Fluentd to send audit logs to Datadog
---

Datadog is a SAAS monitoring and security platform. In this guide, we'll explain

> how to forward Teleport audit events to Datadog, using Fluentd.

Reviewer 2: ```suggestion
how to forward Teleport audit events to Datadog using Fluentd.
```

an API endpoint with keys, so we will use Fluentd as an intermediary.

{/*
The Graph below is made with this mermaidjs logic:
graph LR
style e fill: #a379ff,color: #ffffff
style f fill: #a379ff,color: #ffffff
style c fill: #a379ff,color: #ffffff
style g fill: #a379ff,color: #ffffff
style i fill: #a379ff,color: #ffffff
style e fill: #844cff,color: #ffffff
style monitor fill:#a8afb1,stroke: #000000

a(User)
a-->e

subgraph local [Local Network]
    style local fill:#C0C0C0,stroke: #000000
    direction LR
    subgraph teleport [ ]
        direction TB
        style teleport fill:none,stroke:none
            e(Teleport Proxy)
        e<-->f(Teleport Auth)
        e<-->g(Teleport Service)
        e<-->i(Teleport Service)
        g<-->j([Infrastructure])
        i<-->k([Infrastructure])
    end

    subgraph monitor [Monitoring]
        c(teleport-event-handler)
        d(fluentd)
    end
    c-->d
end
e-->c
d-->h(Datadog)
*/}

![Flowchart of Teleport with events forwarded to Datadog](../../../img/management/datadog-flowchart.png)

## Prereqs

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

- A [Datadog](https://www.datadoghq.com/) account.

> - Fluentd version v(=fluentd.version=) or greater. The Teleport Event Handler

Author: >  Is there a recommended installation method for following this guide?
I went back and forth on how to phrase this in the original writing. I think we want to avoid telling readers the best way to set up 3rd party systems in unknown environments. I'm adding a line that the guide assumes a local docker container for testing.

> Also, do we expect Fluentd to be listening on a specific port for this guide to work?

Yes, and the config file we generate and integrate in the steps below ensures that this port is listening.

  will create a new `fluent.conf` file you can integrate into an existing Fluentd
  system, or use with a fresh setup.

(!docs/pages/includes/tctl.mdx!)

> ## Step 1/6. Install the event handler plugin

Author: ```suggestion
## Step 1/6. Install the Event Handler plugin
```

> ## Step 1/6. Install the event handler plugin

Author: Resolved in a72c62d1ffe6995a82a7dda8a8ee8cfadd846751

The Teleport event handler runs alongside the Fluentd forwarder, receives events
from Teleport's events API, and forwards them to Fluentd.

> (!docs/pages/includes/install-event-handler.mdx!)

Author: Resolving in the partial.

> (!docs/pages/includes/install-event-handler.mdx!)

Author: Resolved in a72c62d1ffe6995a82a7dda8a8ee8cfadd846751

> (!docs/pages/includes/install-event-handler.mdx!)

Author: Updated.

## Step 2/6. Configure the plugin

> (!docs/pages/includes/configure-event-handler.mdx!)

Author: Resolved in a72c62d1ffe6995a82a7dda8a8ee8cfadd846751

## Step 3/6. Create a user and role for reading audit events

The `configure` command generates a file called `teleport-event-handler-role.yaml`
that defines a `teleport-event-handler` role and a user with read-only access
to the `event` API:

```yaml
kind: user
metadata:
  name: teleport-event-handler
spec:
  roles: ['teleport-event-handler']
version: v2
---
kind: role
metadata:
  name: teleport-event-handler
spec:
  allow:
    rules:
      - resources: ['event']
        verbs: ['list','read']
version: v5

> ```

Author: We should probably update all the other guides as well, since this was copied.

  allow:
    impersonate:
      users: ["teleport-event-handler"]
      roles: ["teleport-event-handler"]
```

Next, create the role:

```code
$ tctl create -f teleport-event-handler-impersonator.yaml
```

(!docs/pages/includes/add-role-to-user.mdx role="teleport-event-handler-impersonator"!)

### Export an identity file for the Event Handler plugin user

The Teleport Event Handler plugin uses the `teleport-event-handler` role and user to
read events. We export an identity file for the user with the `tctl auth sign`
command.

<Tabs>
<TabItem label="Self-Hosted" scope={["oss","enterprise"]}>
```code
$ tctl auth sign --user=teleport-event-handler --out=identity
```

This command creates one PEM-encoded file, `identity`. The identity file
includes both TLS and SSH credentials. The Event Handler plugin uses the SSH
credentials to connect to the Proxy Service, which establishes a reverse tunnel
connection to the Auth Service. The plugin uses this reverse tunnel, along with
your TLS credentials, to connect to the Auth Service's gRPC endpoint.

</TabItem>
<TabItem label="Teleport Cloud" scope={["cloud"]}>
```code
$ tctl auth sign --user=teleport-event-handler --out=identity
```

> This command creates one PEM-encoded file, `identity`. The identity file

Reviewer 2: Looks like this entire block is duplicated in both tabs - would it make sense to move it out of the tabs so we don't risk forgetting to update both sections in tandem in the future?

Author: Yes, but can you think of a good title for the consolidated tab label, to identify the tab as referring to Self-Hosted or Cloud, but not Helm", to contrast the "Helm Chart" tab?

Reviewer 2: Hmm. Nothing comes to mind. This is a tricky one. I've already approved though, so if we don't think of a good option feel free to merge as-is.

Author: $5 says @ptgott has got a great idea when he comes around to reviewing this one.

Author: Excellent! Now, who do I collect my $5 from?

Author: Resolved in a72c62d1ffe6995a82a7dda8a8ee8cfadd846751

with the `file` format, then create a secret in Kubernetes.

Create the identity using the following command:

```code
$ tctl auth sign --format=file --user=teleport-event-handler --out=identity
```

Then create the Kubernetes secret:

```code
$ kubectl create secret generic teleport-event-handler-identity --from-file=auth_id=identity
```

These commands should result in a PEM-encoded file, `identity`, and a secret
in Kubernetes with the name `teleport-event-handler-identity`.

</TabItem>
</Tabs>

## Step 5/6. Install Fluentd output plugin for Datadog

In order for Fluentd to communicate with Datadog, it requires the [Fluentd output
plugin for Datadog](https://github.com/DataDog/fluent-plugin-datadog). Install
the plugin on your Fluentd host using either `gem` or the `td-agent`, if installed:

```code
# Using Gem
$ gem install fluent-plugin-datadog

# Using td-agent
$ /usr/sbin/td-agent-gem install fluent-plugin-datadog
```

<Admonition type="tip" title="Testing Locally?">

If you're running Fluentd in a local docker container for testing, you can adjust
the entrypoint to an interactive shell as the root user, so you can install the
plugin before starting Fluentd:

```code

> $ docker run -u $(id -u root):$(id -g root) -p 8888:8888 -v $(pwd):/keys -v \

Reviewer 3: I'm not 100% sure using `id -u root` and `id -g root` is guaranteed to work on all systems. It likely will, but it assumes the ids associated with root on the host system match those on the Docker container.

That being said, I'd think they likely will, so I'll leave it up to you whether this is worth pursuing.

Author: I'd like to put a comment line at the top of the code block about this, but I'm unsure how to phrase it.

$(pwd)/fluent.conf:/fluentd/etc/fluent.conf --entrypoint=/bin/sh -it --tty  fluent/fluentd:edge
# From the container shell:
$ gem install fluent-plugin-datadog
$ fluentd -c /fluentd/etc/fluent.conf
```

</Admonition>

### Configure Fluentd for Datadog

> From the Datadog web UI, generate an API key for Fluentd. From **Organization Settings** -> Access -> API Keys, click on **+ New Key**:

Author: Not all, since "Access" is a header and not a link, but I will update "API Keys".

> From the Datadog web UI, generate an API key for Fluentd. From **Organization Settings** -> Access -> API Keys, click on **+ New Key**:

Author: I didn't want to add a link since it will be different depending on your colo. I navigated to it from the user menu:

![image](https://user-images.githubusercontent.com/2349184/221257575-9d052504-6aa3-48c6-88c0-b3735d00a978.png)

But is "User Menu" the right generic term for that?

![Datadog API key generation](../../../img/management/datadog-generate-api.png)

Copy the API key, and use it to add a new `<match>` block to `fluent.conf`:

> 

Reviewer: Might be worth noting that the `ca_path`, `cert_path` and `private_key_path` need to match the paths to the credentials you generated in `fluent.conf`.

Author: Resolved in a72c62d1ffe6995a82a7dda8a8ee8cfadd846751

  --values teleport-plugin-event-handler-values.yaml \
  --version (=teleport.plugin.version=)
```

</TabItem>
</Tabs>

<Admonition type="note">

This example will start exporting from `May 5th 2021`:

```code
$ ./teleport-event-handler start --config teleport-event-handler.toml --start-time "2022-02-02T00:00:00Z"
```

The start time can be set only once, on the first run of the tool.

If you want to change the time frame later, remove the plugin state directory
that you specified in the `storage` field of the handler's configuration file.

</Admonition>

Once the handler starts, you will see notifications in Fluentd about scanned and forwarded events:

```txt
INFO[0046] Event sent id=0b5f2a3e-faa5-4d77-ab6e-362bca0994fc ts="2021-06-08 11:00:56.034 +0000 UTC" type=user.login
...
```

The Logs view in Datadog should now report your Teleport cluster events:

![Datadog Logs](../../../img/management/datadog-logs.png)

Review the Fluentd output plugin for Datadog [README
file](https://github.com/DataDog/fluent-plugin-datadog/blob/master/README.md)
to learn how to customize the log format entering Datadog.

> 

Reviewer: It might be good to add a quick section about visualizing audit event counts by the "event" field, e.g., in the Splunk guide (https://goteleport.com/docs/management/export-audit-events/splunk/?scope=enterprise#step-44-visualize-your-audit-logs-in-splunk)

Author: It might be good, but it looks like that would require extra configuration to format the data so it could be separated by type, and I'd rather not go that deep into documenting a third party plugin written for another third party tool.

Reviewer: As far as I know, Datadog should be automatically parsing the JSON documents the plugin is sending to it—is that not the case in this setup?

Author: Nope, doesn't seem to be.

## Troubleshooting connection issues

> 

Reviewer: I got "connection reset by peer" errors when running the Event Handler:

```
ERRO   Error sending event to Teleport: Post "https://localhost:8888/test.log": read tcp 127.0.0.1:39460->127.0.0.1:8888: read: connection reset by peer event-handler/app.go:118
```

Looks like this has to do with the TLS handshake between the Event Handler and Fluentd, as reproduced via `curl`:

```
$ curl --key client.key --cert client.crt --cacert ca.crt -v https://localhost:8888/test.log
*   Trying 127.0.0.1:8888...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8888 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: ca.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* OpenSSL SSL_connect: Connection reset by peer in connection to localhost:8888
* Closing connection 0
curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to localhost:8888
```

Weirdly, the client cert's `Not After` is before its `Not Before`:

```
~$ openssl x509 -text -in client.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            8a:82:e4:d5:f1:7b:c4:e8:a9:56:8e:ee:64:7f:87:7c
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = US, CN = localhost
        Validity
            Not Before: Feb 23 23:22:46 2023 GMT
            Not After : Feb 20 23:22:46 2033 GMT
```

I built the Event Handler from commit `bc85392dc3b03bedcd963709d3ef4bba2af37b2d`.

Author: That's from [this issue](https://github.com/gravitational/teleport-plugins/issues/747). I'm hoping it will be resolved soon, since it affects all guides using the generated configuration.
